crates/ruff_python_formatter/src/statement/suite.rs

```diff
@@ -47,164 +51,201 @@ impl FormatRule<Suite, PyFormatContext<'_>> for FormatSuite {
         let comments = f.context().comments().clone();
         let source = f.context().source();
 
-        let mut iter = statements.iter();
-        let Some(first) = iter.next() else {
-            return Ok(());
-        };
-
         let mut f = WithNodeLevel::new(node_level, f);
+        write!(
+            f,
+            [format_with(|f| {
+                let mut iter = statements.iter();
+                let Some(first) = iter.next() else {
+                    return Ok(());
+                };
 
-        // Format the first statement in the body, which often has special formatting rules.
-        let mut last = first;
-        match self.kind {
-            SuiteKind::Other => {
-                if is_class_or_function_definition(first) && !comments.has_leading_comments(first) {
-                    // Add an empty line for any nested functions or classes defined within
-                    // non-function or class compound statements, e.g., this is stable formatting:
-                    // ```python
-                    // if True:
-                    //
-                    //     def test():
-                    //         ...
-                    // ```
-                    write!(f, [empty_line()])?;
-                }
-                write!(f, [first.format()])?;
-            }
-            SuiteKind::Class if is_docstring(first) => {
-                if !comments.has_leading_comments(first) && lines_before(first.start(), source) > 1
-                {
-                    // Allow up to one empty line before a class docstring, e.g., this is
-                    // stable formatting:
-                    // ```python
-                    // class Test:
-                    //
-                    //     """Docstring"""
-                    // ```
-                    write!(f, [empty_line()])?;
-                }
-                write!(f, [first.format()])?;
-
-                // Enforce an empty line after a class docstring, e.g., these are both stable
-                // formatting:
-                // ```python
-                // class Test:
-                //     """Docstring"""
-                //
-                //     ...
-                //
-                //
-                // class Test:
-                //
-                //     """Docstring"""
-                //
-                //     ...
-                // ```
-                if let Some(second) = iter.next() {
-                    // Format the subsequent statement immediately. This rule takes precedence
-                    // over the rules in the loop below (and most of them won't apply anyway,
-                    // e.g., we know the first statement isn't an import).
-                    write!(f, [empty_line(), second.format()])?;
-                    last = second;
```

Comment:
I had to move this into the `loop to avoid having to handle the suppression comments on `second`